### DISCUSSION TITLE
CTT Attributes Write Value Test fails for null Values

### DISCUSSION THREAD

**Discussion by johannwesely** [eclipse-milo/milo#1298] [General]
Hi, im currently testing our Milo Server with CTT now in the Attribute Write Value test i have lots of failures. 
![cttattributewrite](https://github.com/user-attachments/assets/66a3632b-e759-4146-afcd-d3fbd73f774c)
I figured out using wireshark this happens  because our Server has initaly null Values set, CTT reads the values and wants to write them back and gets BadTypeMismatch. Now i checked in Milo Dev1.0 why this happens and came to the Solution that in Server UaVariableNode Line 250 the server Throws the BadTypeMismatch when the value to write is null.  I changed the Line to  
 ` Object o = variant.getValue();
    if (o == null) return value;`
 Now i have a lot More green but still some Failures left do you think this could be a solution.
![cttattributewrite2](https://github.com/user-attachments/assets/3f98e596-67bc-43ed-a3c8-b2447b0324c3)
 I also tried to initalize the Values before Starting CTT Testscripts and this works also. But can you tell me if null Values should be possible to write ? I cannot find a mention in OpcUa Spec?

**Comment by kevinherron** [comment]
Whether or not a server allows null values to be written is up to the implementation. The way the SDK is currently written does not support it. I think originally I didn't support it at the attribute level in the UaNodes either, but that seems to have been relaxed a bit.

You should just provide valid initial values for the purpose of CTT testing and chase remaining legitimate errors.

**Reply by johannwesely** [reply]
Ah okay I liked the idea of null Values for Inital Values. Because later I dont know if our Process has written for example an 0 value or if it is still the inital 0. But Thanks for your Information saves me some time.

**Reply by kevinherron** [reply]
You can use `null` or 0 + `Uncertain_InitialValue` StatusCode in that situation.

Null values work internally, they just don't work when written by an external client.

There's also certain attributes that aren't allowed to have `null` values (any attribute that is a structure).

And there's some weird behavior you'll see with optional attributes, because when reading an optional attribute `null` is used to decide `Bad_AttributeIdInvalid` should be returned because it's not supported/implemented.
